base/complex.jl

```diff
@@ -514,13 +514,13 @@ function robust_cinv(c::Float64, d::Float64)
 end
 
 function ssqs(x::T, y::T) where T<:Real
-    k::Int = 0
+    k = 0
     ρ = x*x + y*y
     if !isfinite(ρ) && (isinf(x) || isinf(y))
         ρ = convert(T, Inf)
     elseif isinf(ρ) || (ρ==0 && (x!=0 || y!=0)) || ρ<nextfloat(zero(T))/(2*eps(T)^2)
         m::T = max(abs(x), abs(y))
-        k = m==0 ? m : exponent(m)
+        k = m==0 ? 0 : convert(Int, exponent(m))
         xk, yk = ldexp(x,-k), ldexp(y,-k)
         ρ = xk*xk + yk*yk
```

Comment:
Unrelated to the issue you're trying to solve here, so feel free to ignore, but while at it: For `k==0`, it seems this doesn't change anything, and as we're checking `m==0`  anyway (which implies `k==0`), maybe this could be slightly more efficient:
```suggestion
    elseif isinf(ρ) || (ρ==0 && (x!=0 || y!=0)) || ρ<nextfloat(zero(T))/(2*eps(T)^2)
        m::T = max(abs(x), abs(y))
        if m != 0
            k = convert(Int, exponent(m))
            xk, yk = ldexp(x,-k), ldexp(y,-k)
            ρ = xk*xk + yk*yk
        end
```
But at that point, it might make sense to subsume this in the condition above. Doesn't `m == 0` imply `x==0 && y==0` or is there some sneaky floating point thing going on so that `abs(x)==0` even if `x!=0`? Note further that that condition is a bit strange to begin with: Assuming `eps(T)<sqrt(0.5)` for any reasonable float, `nextfloat(zero(T))/(2*eps(T)^2)` will be be greater than `0`. So for `ρ==0`, the condition will always be true, so matter `x!=0 || y!=0`. However, if we want to exclude `x==0 && y==0`, we could actually rewrite this as
```suggestion
    elseif isinf(ρ) || (ρ<nextfloat(zero(T))/(2*eps(T)^2) && (x!=0 || y!=0))
        m::T = max(abs(x), abs(y))
        k = convert(Int, exponent(m))
        xk, yk = ldexp(x,-k), ldexp(y,-k)
        ρ = xk*xk + yk*yk
```

Reply:
That sounds reasonable to me, I'm happy to change it to that.

A slight optimization on your version could be:
```julia
    elseif isinf(ρ) || ((x!=0 || y!=0) && ρ<nextfloat(zero(T))/(2*eps(T)^2))
```
since presumably comparisons to 0 are faster than those arithmetic operations, and a more common situation is that this function is being called with `x==0` and `y==0` so in that common case this will skip the check on `ρ`.

From what I can tell this code path only gets called when `x` and/or `y` are very small or large but still finite and one or the other is nonzero. Here's a demonstration that in some limiting cases your suggestion is the same as the old code logic:
```julia
function ssqs1(x::T, y::T) where T<:Real
    k = 0
    ρ = x*x + y*y
    if !isfinite(ρ) && (isinf(x) || isinf(y))
        ρ = convert(T, Inf)
    elseif isinf(ρ) || (ρ==0 && (x!=0 || y!=0)) || ρ<nextfloat(zero(T))/(2*eps(T)^2)
        m::T = max(abs(x), abs(y))
        k = m==0 ? 0 : convert(Int, exponent(m))
        xk, yk = ldexp(x,-k), ldexp(y,-k)
        ρ = xk*xk + yk*yk
    end
    ρ, k
end

function ssqs2(x::T, y::T) where T<:Real
    k = 0
    ρ = x*x + y*y
    if !isfinite(ρ) && (isinf(x) || isinf(y))
        ρ = convert(T, Inf)
    elseif isinf(ρ) || ((x!=0 || y!=0) && ρ<nextfloat(zero(T))/(2*eps(T)^2))
        m::T = max(abs(x), abs(y))
        k = convert(Int, exponent(m))
        xk, yk = ldexp(x,-k), ldexp(y,-k)
        ρ = xk*xk + yk*yk
    end
    ρ, k
end

for (x, y) in (
  (1e-146, 0.0),
  (1e-147, 0.0),
  (1e-161, 0.0),
  (1e-162, 0.0),
  (1e154, 0.0),
  (1e155, 0.0),
)
  @show x, y
  @show ρ = x*x + y*y
  @show ssqs1(x, y)
  @show ssqs2(x, y)
  println()
end
```
which outputs:
```julia
(x, y) = (1.0e-146, 0.0)
ρ = x * x + y * y = 1.0e-292
ssqs1(x, y) = (1.0e-292, 0)
ssqs2(x, y) = (1.0e-292, 0)

(x, y) = (1.0e-147, 0.0)
ρ = x * x + y * y = 1.0e-294
ssqs1(x, y) = (2.5546755962044414, -489)
ssqs2(x, y) = (2.5546755962044414, -489)

(x, y) = (1.0e-161, 0.0)
ρ = x * x + y * y = 1.0e-322
ssqs1(x, y) = (1.2650140831706915, -535)
ssqs2(x, y) = (1.2650140831706915, -535)

(x, y) = (1.0e-162, 0.0)
ρ = x * x + y * y = 0.0
ssqs1(x, y) = (3.2384360529169696, -539)
ssqs2(x, y) = (3.2384360529169696, -539)

(x, y) = (1.0e154, 0.0)
ρ = x * x + y * y = 1.0e308
ssqs1(x, y) = (1.0e308, 0)
ssqs2(x, y) = (1.0e308, 0)

(x, y) = (1.0e155, 0.0)
ρ = x * x + y * y = Inf
ssqs1(x, y) = (3.476677903917502, 514)
ssqs2(x, y) = (3.476677903917502, 514)
```

Reply:
I agree with your assessment that I don't see how `m` could be zero inside this branch of the code logic so I don't know why that was being checked in the first place.

Reply:
With the old logic it could, as with `y==0 && y==0`, clearly `(ρ==0 && (x!=0 || y!=0))` would be false, but `ρ<nextfloat(zero(T))/(2*eps(T)^2)` would be true (and `m==0`, obviously). So I guess that check on `m` was necessary because the branch condition did not do what it was intended to.